### PR TITLE
🐛Make the regex less restrictive to allow underscore etc. for ssh key file name

### LINKED ISSUE #2060: ssh key validation regex is too restrictive
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]

created an AWSMachine with an ssh key with an underscore in the name, got denied by webhook.

**What did you expect to happen:**

it should work, as I believe all ascii characters should work for ssh key name:

https://docs.aws.amazon.com/cli/latest/reference/ec2/create-key-pair.html

**Environment:**

- Cluster-api-provider-aws version: 0.6.1

### PR DESCRIPTION
**Pull request by arghya88**
**What this PR does / why we need it**:
Make the regex less restrictive to allows underscore etc. for ssh key file name
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2060

### CODE REVIEW

**Comment by arghya88** [comment]
cc @randomvariable

File: api/v1alpha3/webhooks.go [hunk 1]
```diff
@@ -38,7 +38,7 @@ func aggregateObjErrors(gk schema.GroupKind, name string, allErrs field.ErrorLis
 func isValidSSHKey(sshKey *string) field.ErrorList {
 	var allErrs field.ErrorList
 	if sshKey != nil {
-		reg, err := regexp.Compile("[^-A-Za-z0-9-]+")
+		reg, err := regexp.Compile("[^-A-Za-z0-9_-]+")
```
Comment (randomvariable) [inline api/v1alpha3/webhooks.go:41]:
```suggestion
		reg, err := regexp.Compile("[\x00-\x7F]+")
```

Also, can we move this into variable using `regexp.MustCompile`, so we're not reallocating memory on every call to the webhook?

Comment (arghya88) [reply api/v1alpha3/webhooks.go:41]: If I do `reg := regexp.MustCompile(`[\x00-\x7F]+`)` the tests with ssh key name `test-key`, `test_key`, `test` fails
`

Comment (arghya88) [reply api/v1alpha3/webhooks.go:41]: I have done some changes. with this change `test\t` is a valid name of ssh keyfile.Hence had to make some changes in the test as well. PTAL

File: api/v1alpha3/webhooks.go [hunk 2]
```diff
@@ -38,10 +38,7 @@ func aggregateObjErrors(gk schema.GroupKind, name string, allErrs field.ErrorLis
 func isValidSSHKey(sshKey *string) field.ErrorList {
 	var allErrs field.ErrorList
 	if sshKey != nil {
-		reg, err := regexp.Compile("[^-A-Za-z0-9-]+")
-		if err != nil {
-			return append(allErrs, field.Invalid(field.NewPath("sshKey"), sshKey, "SSHKey contains invalid character"))
-		}
+		reg := regexp.MustCompile(`^\x00-\x7F]+`)
```
Comment (randomvariable) [inline api/v1alpha3/webhooks.go:41]: Please move this to the package variable as mentioned previously.

Comment (randomvariable) [inline api/v1alpha3/webhooks.go:41]: Let's make sure we get this right:

Valid SSH key names are the intersection of ISO-8859, ASCII and valid XML entity names.

That gives us a regex of 

```go
var (
  sshKeyValidNameRegex regexp.MustCompile(`^[[:graph:]]+([[:print:]]*[[:graph:]]+)*$`)
)`
```

This should go in a package variable, such that the controller panics as close as possible to startup if the regex is invalid, and that we minimise memory reallocations.

then we can do something like the following (assuming we can change isValidSSHKey's function return  signature from `field.ErrorList` to `error`.

```suggestion
		if sshKey != nil && !sshKeyValidNameRegex.Match([]byte(*sshKey))
		{
		  return field.Invalid(field.NewPath("sshKey"), sshKey, "Name is invalid. Must be specified in ASCII and must not start or end in whitespace")
		}
		return nil
```

Comment (arghya88) [reply api/v1alpha3/webhooks.go]: Make sense..I think the change for error handling needs to be done in a separate PR

Comment (randomvariable) [reply api/v1alpha3/webhooks.go]: Yup, that's fine.

**Review by randomvariable — APPROVED** [review]

**Pull request merged**